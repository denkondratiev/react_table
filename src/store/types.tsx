import {
  setParams,
  setTable,
  setRows,
  setCells,
  setShowButtons,
  increment,
  setNewRow,
  removeRow
} from './actions'

export const ACTION_TYPES = {
  SET_PARAMS: 'SET::PARAMS',
  SET_TABLE: 'SET::TABLE',
  SET_ROWS: 'SET::ROWS',
  SET_CELLS: 'SET::CELLS',
  SHOW_BUTTONS: 'SHOW::BUTTONS',
  INCREMENT: 'INCREMENT',
  ADD_ROW: 'ADD::ROW',
  REMOVE_ROW: 'REMOVE::ROW'
}

export type State = {
  table: Array<string>
  rows: RowsParams
  cells: CellsParams
  params: Params
  buttons: boolean
}

export type Action = (
  | ReturnType<typeof setParams>
  | ReturnType<typeof setTable>
  | ReturnType<typeof setRows>
  | ReturnType<typeof setCells>
  | ReturnType<typeof setShowButtons>
  | ReturnType<typeof increment>
  | ReturnType<typeof setNewRow>
  | ReturnType<typeof removeRow>
)

export type Params = {
  [name: string]: string | number
}

export type RowsParams = {
  [name: string]: Array<string>
}

export type CellsParams = {
  [name: string]: { id: string, amount: number }
}

export type NewRowsParams = {
  table: Array<string>
  rows: RowsParams
  cells: CellsParams
}
