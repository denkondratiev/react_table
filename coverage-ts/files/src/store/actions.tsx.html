
  <!DOCTYPE html>
  <html>
    <head>
      <title>actions.tsx</title>
      <link href="https://cdn.jsdelivr.net/npm/semantic-ui@2.4.2/dist/semantic.min.css" type="text/css" rel="stylesheet">
      <script src="https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.52.2/codemirror.min.js" type="text/javascript" charset="utf-8"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.52.2/mode/javascript/javascript.min.js" type="text/javascript" charset="utf-8"></script>
<link href="https://cdnjs.cloudflare.com/ajax/libs/codemirror/5.52.2/codemirror.min.css" type="text/css" rel="stylesheet">
<script src="..\..\..\assets\source-file.js" type="text/javascript" charset="utf-8"></script>
<link href="..\..\..\assets\source-file.css" type="text/css" rel="stylesheet">
    </head>
    <body>
    <div style="margin-top:3em" class="ui container"><h1 class="ui header"><a href="..\..\..\index.html">TypeScript coverage report</a></h1><table style="margin-top:2em" class="ui celled table"><thead class=""><tr class=""><th class="">Filename</th><th class="">Percent</th><th class="">Threshold</th><th class="">Total</th><th class="">Covered</th><th class="">Uncovered</th></tr></thead><tbody class=""><tr class="positive"><td class="">src\store\actions.tsx</td><td class="">97.35%</td><td class="">80%</td><td class="">113</td><td class="">110</td><td class="">3</td></tr></tbody></table><textarea id="editor" readonly="" style="margin-top:3em">export const ACTION_TYPES = {
  SET_PARAMS: &#x27;SET::PARAMS&#x27;,
  SET_TABLE: &#x27;SET::TABLE&#x27;,
  SET_ROWS: &#x27;SET::ROWS&#x27;,
  SET_CELLS: &#x27;SET::CELLS&#x27;,
  SHOW_BUTTONS: &#x27;SHOW::BUTTONS&#x27;,
  INCREMENT: &#x27;INCREMENT&#x27;,
  ADD_ROW: &#x27;ADD::ROW&#x27;,
  REMOVE_ROW: &#x27;REMOVE::ROW&#x27;
}

// TYPES

export type State = {
  table: Array&lt;string&gt;
  rows: RowsParams
  cells: CellsParams
  params: Params
  buttons: boolean
}

export type Action = (
  | ReturnType&lt;typeof setParams&gt;
  | ReturnType&lt;typeof setTable&gt;
  | ReturnType&lt;typeof setRows&gt;
  | ReturnType&lt;typeof setCells&gt;
  | ReturnType&lt;typeof setShowButtons&gt;
  | ReturnType&lt;typeof increment&gt;
  | ReturnType&lt;typeof setNewRow&gt;
  | ReturnType&lt;typeof removeRow&gt;
)

export type Params = {
  [name: string]: string | number
}

export type RowsParams = {
  [name: string]: Array&lt;string&gt;
}

export type CellsParams = {
  [name: string]: { id: string, amount: number }
}


export type NewRowsParams = {
  table: Array&lt;string&gt;
  rows: RowsParams
  cells: CellsParams
}

// ACTIONS

export const setParams = (params: Params) =&gt; ({
  payload: params,
  type: ACTION_TYPES.SET_PARAMS
})

export const setTable = (params: Array&lt;string&gt;) =&gt; ({
  payload: params,
  type: ACTION_TYPES.SET_TABLE
})

export const setRows = (params: RowsParams) =&gt; ({
  payload: params,
  type: ACTION_TYPES.SET_ROWS
})

export const setCells = (params: CellsParams) =&gt; ({
  payload: params,
  type: ACTION_TYPES.SET_CELLS
})

export const setShowButtons = (params: boolean) =&gt; ({
  payload: params,
  type: ACTION_TYPES.SHOW_BUTTONS
})

export const increment = (params: string) =&gt; ({
  payload: params,
  type: ACTION_TYPES.INCREMENT
})

export const setNewRow = (params: NewRowsParams) =&gt; ({
  payload: params,
  type: ACTION_TYPES.ADD_ROW
})

export const removeRow = (params: any) =&gt; ({
  payload: params,
  type: ACTION_TYPES.REMOVE_ROW
})
</textarea><pre id="annotations" style="display:none">[{&quot;file&quot;:&quot;src\\store\\actions.tsx&quot;,&quot;line&quot;:88,&quot;character&quot;:26,&quot;text&quot;:&quot;params&quot;},{&quot;file&quot;:&quot;src\\store\\actions.tsx&quot;,&quot;line&quot;:89,&quot;character&quot;:2,&quot;text&quot;:&quot;payload&quot;},{&quot;file&quot;:&quot;src\\store\\actions.tsx&quot;,&quot;line&quot;:89,&quot;character&quot;:11,&quot;text&quot;:&quot;params&quot;}]</pre></div>
    <p class="footer-text">TypeScript Coverage Report generated by <a href="https://github.com/plantain-00/type-coverage">type-coverage</a> and <a href="https://github.com/alexcanessa/typescript-coverage-report">typescript-coverage-report</a> at Thu, 30 Jul 2020 09:48:47 GMT</p>
    </body>
  </html>
  