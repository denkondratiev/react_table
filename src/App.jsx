import React from 'react'
import Form from './components/Form/Form'
import Table from './components/Table/Table'
import Buttons from './components/Buttons/Buttons'
import './App.css'

const App = () => (
  <div className="main">
    <Form />
    <Buttons />
    <Table/>
  </div>
)

export default App
