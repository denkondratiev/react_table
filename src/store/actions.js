export const ACTION_TYPES = {
  SET_COLUMNS: 'SET::COLUMNS',
  SET_ROWS: 'SET::ROWS',
  SET_TABLE: 'SET::TABLE',
  SET_HIGHLIGHT: 'SET::HIGHLIGHT',
  SET_SHOW_TABLE: 'SET::SHOW::TABLE',
  SET_SHOW_BUTTONS: 'SET::SHOW::BUTTONS'
}

export const setColumns = (param) => ({
  payload: param === '' ? '' : Number(param),
  type: ACTION_TYPES.SET_COLUMNS
})

export const setRows = (param) => ({
  payload: param === '' ? '' : Number(param),
  type: ACTION_TYPES.SET_ROWS
})

export const setHighlights = (param) => ({
  payload: param === '' ? '' : Number(param),
  type: ACTION_TYPES.SET_HIGHLIGHT
})

export const setTable = (param) => ({
  payload: param,
  type: ACTION_TYPES.SET_TABLE
})

export const setShowTable = (param) => ({
  payload: param,
  type: ACTION_TYPES.SET_SHOW_TABLE
})

export const setShowButtons = (param) => ({
  payload: param,
  type: ACTION_TYPES.SET_SHOW_BUTTONS
})
