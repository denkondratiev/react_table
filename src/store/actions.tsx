import {
  ACTION_TYPES,
  Params,
  RowsParams,
  CellsParams,
  NewRowsParams
} from './types'

export const setParams = (params: Params) => ({
  payload: params,
  type: ACTION_TYPES.SET_PARAMS
})

export const setTable = (params: Array<string>) => ({
  payload: params,
  type: ACTION_TYPES.SET_TABLE
})

export const setRows = (params: RowsParams) => ({
  payload: params,
  type: ACTION_TYPES.SET_ROWS
})

export const setCells = (params: CellsParams) => ({
  payload: params,
  type: ACTION_TYPES.SET_CELLS
})

export const setShowButtons = (params: boolean) => ({
  payload: params,
  type: ACTION_TYPES.SHOW_BUTTONS
})

export const increment = (params: string) => ({
  payload: params,
  type: ACTION_TYPES.INCREMENT
})

export const setNewRow = (params: NewRowsParams) => ({
  payload: params,
  type: ACTION_TYPES.ADD_ROW
})

export const removeRow = (params: any) => ({
  payload: params,
  type: ACTION_TYPES.REMOVE_ROW
})
