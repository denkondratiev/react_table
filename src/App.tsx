import React from 'react'
import Form from './components/Form/Form'
import Table from './components/Table/Table'
import Buttons from './components/Buttons/Buttons'
import './App.css'

const App: React.FC = () => (
  <div className="main">
    <Form />
    <Buttons />
    <Table />
  </div>
)

export default App
