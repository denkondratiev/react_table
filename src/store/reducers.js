import { createStore } from 'redux'
import { ACTION_TYPES } from './actions'

function rootReducer (state, action) {
  const {
    SET_COLUMNS,
    SET_ROWS,
    SET_HIGHLIGHTS,
    SET_TABLE,
    SET_SHOW_TABLE,
    SET_SHOW_BUTTONS,
    INCREMENT,
    SET_LIGHT_CELLS
  } = ACTION_TYPES

  switch (action.type) {
    case SET_TABLE: {
      return {
        ...state,
        table: action.payload
      }
    }

    case SET_COLUMNS: {
      return {
        ...state,
        columns: action.payload === '' ? '' : Number(action.payload)
      }
    }

    case SET_ROWS: {
      return {
        ...state,
        rows: action.payload === '' ? '' : Number(action.payload)
      }
    }

    case SET_HIGHLIGHTS: {
      return {
        ...state,
        highlights: action.payload === '' ? '' : Number(action.payload)
      }
    }

    case SET_SHOW_TABLE: {
      return {
        ...state,
        showTable: action.payload
      }
    }

    case SET_SHOW_BUTTONS: {
      return {
        ...state,
        showButtons: action.payload
      }
    }

    case INCREMENT: {
      state.table.splice(action.payload.id, 1, [...state.table[action.payload.id]
        .map((item, index) => index !== action.payload.cellIndex
          ? item
          : {
            ...state.table[action.payload.id][action.payload.cellIndex],
            amount: state.table[action.payload.id][action.payload.cellIndex].amount + 1
          })
      ])
      const newTable = [...state.table]
      return {
        ...state,
        table: newTable
      }
    }

    case SET_LIGHT_CELLS: {
      // const amount = action.payload.amount
      // const highlights = action.payload.highlights
      const rowId = action.payload.id
      const cellIndex = action.payload.cellIndex
      // const col = state.table.flat().length / state.table.length
      console.log(rowId, cellIndex)
      break

      // const lightArray = [...state.table].flat().map((item, index) => ({
      //   ...item,
      //   rowIndex: index,
      //   difference: Math.abs(item.amount - amount)
      // })).sort((a, b) => a.difference - b.difference).slice(0, highlights + 1)

      // state.table.map(cell => (
      //   lightArray.some(lightCell => lightCell.cellId === cell.cellId)
      //     ? { ...cell, isLight: true }
      //     : cell
      // ))

      // return {
      //   ...state,
      //   table: newTable
      // }
    }

    default:
      return state
  }
}

const initialState = {
  rows: '',
  columns: '',
  highlights: '',
  table: [],
  showTable: false,
  showButtons: false
}

const store = createStore(
  rootReducer,
  initialState,
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
)

export default store
