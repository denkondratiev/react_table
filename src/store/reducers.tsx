import { combineReducers } from 'redux'
import { ACTION_TYPES } from './actions'
import { Action, Params, RowsParams, CellsParams } from './actions'

const paramsReducer = (state: Params = {}, action: Action): Params => {
  const { SET_PARAMS } = ACTION_TYPES

  switch (action.type) {
    case SET_PARAMS:
      return action.payload
    default:
      return state
  }
}

function tableReducer (state: Array<string> = [], action: Action): Array<string> {
  const { SET_TABLE, ADD_ROW, REMOVE_ROW } = ACTION_TYPES

  switch (action.type) {
    case SET_TABLE:
      return action.payload
    case ADD_ROW:
      return [...state, action.payload.table]
    case REMOVE_ROW:
      return state.filter(id => id !== action.payload.lastRow)
    default:
      return state
  }
}

const rowsReducer = (state: RowsParams = {}, action: Action): RowsParams => {
  const { SET_ROWS, ADD_ROW, REMOVE_ROW } = ACTION_TYPES

  switch (action.type) {
    case SET_ROWS:
      return action.payload
    case ADD_ROW:
      return { ...state, ...action.payload.rows }
    case REMOVE_ROW:
      delete state[action.payload.lastRow]
      return state
    default:
      return state
  }
}


const cellsReducer = (state: CellsParams = {}, action: Action): CellsParams => {
  const { SET_CELLS, INCREMENT, ADD_ROW } = ACTION_TYPES

  switch (action.type) {
    case SET_CELLS:
      return action.payload
    case ADD_ROW:
      return { ...state, ...action.payload.cells }
    case INCREMENT:
      return {
        ...state,
        [action.payload]: {
          ...state[action.payload],
          amount: state[action.payload].amount + 1
        }
      }
    default:
      return state
  }
}

const buttonsReducer = (state: boolean = false, action: Action): boolean => {
  const { SHOW_BUTTONS } = ACTION_TYPES

  switch (action.type) {
    case SHOW_BUTTONS:
      return action.payload
    default:
      return state
  }
}

export const rootReducer = combineReducers({
  params: paramsReducer,
  table: tableReducer,
  rows: rowsReducer,
  cells: cellsReducer,
  buttons: buttonsReducer
})
