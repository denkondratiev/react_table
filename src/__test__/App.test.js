import React from 'react'
import { Provider } from 'react-redux'
import App from '../App'
import { act, create } from 'react-test-renderer'
import { createStore } from 'redux'
import { rootReducer } from '../store/reducers'

// START STORE
const params = { rowsAmount: 2, columnsAmount: 2, lightsAmount: 2 }
const table = ['row1', 'row2']
const rows = {
  row1: ['somecell1', 'somecell2'],
  row2: ['somecell3', 'somecell4']
}

const cells = {
  somecell1: { id: 'somecell1', amount: 1 },
  somecell2: { id: 'somecell2', amount: 2 },
  somecell3: { id: 'somecell3', amount: 0 },
  somecell4: { id: 'somecell4', amount: 1 }
}
const buttons = true
const store = createStore(rootReducer, { params, table, rows, cells, buttons })
// END STORE

it('should correct render with some values', () => {
  const tree = create(
    <Provider store={store}>
      <App />
    </Provider>
  )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

describe('test cases for buttons', () => {
  let component
  let root

  beforeEach(() => {
    act(() => {
      component = create(
        <Provider store={store}>
          <App />
        </Provider>
      )
    })
    root = component.root
  })

  it('click on button delete', () => {
    const id = 'button-delete'

    const ButtonDelete = root.findByProps({ id })

    act(() => ButtonDelete.props.onClick())

    expect(component.toJSON()).toMatchSnapshot()
  })

  // it('click on button add', () => {
  //   const id = 'button-add'

  //   const ButtonAdd = root.findByProps({ id })

  //   act(() => ButtonAdd.props.onClick())

  //   expect(component.toJSON()).toMatchSnapshot()
  // })
})
