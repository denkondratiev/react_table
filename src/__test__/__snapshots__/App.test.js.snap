// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correct render with some values 1`] = `
<div
  className="main"
>
  <form
    className="form"
    id="form"
    onSubmit={[Function]}
  >
    <input
      className="form-control false"
      name="rowsAmount"
      onChange={[Function]}
      placeholder="Rows..."
      type="number"
      value=""
    />
    <input
      className="form-control false"
      name="columnsAmount"
      onChange={[Function]}
      placeholder="Columns..."
      type="number"
      value=""
    />
    <input
      className="form-control false"
      name="lightsAmount"
      onChange={[Function]}
      placeholder="Highlight cells..."
      type="number"
      value=""
    />
    <button
      className="btn btn-dark"
      id="button-generate"
      type="submit"
    >
      Generate
    </button>
  </form>
  <div
    className="buttons"
  >
    <button
      className="btn btn-dark"
      id="button-delete"
      onClick={[Function]}
      type="button"
    >
      Delete row -
    </button>
    <button
      className="btn btn-dark"
      id="button-add"
      onClick={[Function]}
      type="button"
    >
      Add row +
    </button>
  </div>
  <table
    className="table table-dark"
  >
    <tbody>
      <tr
        id="row1"
      >
        <td
          id="somecell1"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={
            Object {
              "background": "",
            }
          }
        >
          1
        </td>
        <td
          id="somecell2"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={
            Object {
              "background": "",
            }
          }
        >
          2
        </td>
        <td
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          Sum: 
          3
        </td>
      </tr>
      <tr
        id="row2"
      >
        <td
          id="somecell3"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={
            Object {
              "background": "",
            }
          }
        >
          0
        </td>
        <td
          id="somecell4"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={
            Object {
              "background": "",
            }
          }
        >
          1
        </td>
        <td
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          Sum: 
          1
        </td>
      </tr>
      <tr>
        <td>
          Average sum: 
          0
        </td>
        <td>
          Average sum: 
          1
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`test cases for buttons click on button delete 1`] = `
<div
  className="main"
>
  <form
    className="form"
    id="form"
    onSubmit={[Function]}
  >
    <input
      className="form-control false"
      name="rowsAmount"
      onChange={[Function]}
      placeholder="Rows..."
      type="number"
      value=""
    />
    <input
      className="form-control false"
      name="columnsAmount"
      onChange={[Function]}
      placeholder="Columns..."
      type="number"
      value=""
    />
    <input
      className="form-control false"
      name="lightsAmount"
      onChange={[Function]}
      placeholder="Highlight cells..."
      type="number"
      value=""
    />
    <button
      className="btn btn-dark"
      id="button-generate"
      type="submit"
    >
      Generate
    </button>
  </form>
  <div
    className="buttons"
  >
    <button
      className="btn btn-dark"
      id="button-delete"
      onClick={[Function]}
      type="button"
    >
      Delete row -
    </button>
    <button
      className="btn btn-dark"
      id="button-add"
      onClick={[Function]}
      type="button"
    >
      Add row +
    </button>
  </div>
  <table
    className="table table-dark"
  >
    <tbody>
      <tr
        id="row1"
      >
        <td
          id="somecell1"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={
            Object {
              "background": "",
            }
          }
        >
          1
        </td>
        <td
          id="somecell2"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={
            Object {
              "background": "",
            }
          }
        >
          2
        </td>
        <td
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          Sum: 
          3
        </td>
      </tr>
      <tr>
        <td>
          Average sum: 
          1
        </td>
        <td>
          Average sum: 
          2
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
